File: android-app/src/main/kotlin/com/jworks/kanjiquest/android/ui/flashcard/FlashcardStudyViewModel.kt
package com.jworks.kanjiquest.android.ui.flashcard

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jworks.kanjiquest.core.domain.model.Kanji
import com.jworks.kanjiquest.core.domain.model.SrsCard
import com.jworks.kanjiquest.core.domain.model.Vocabulary
import com.jworks.kanjiquest.core.domain.repository.FlashcardRepository
import com.jworks.kanjiquest.core.domain.repository.KanjiRepository
import com.jworks.kanjiquest.core.domain.repository.SrsRepository
import com.jworks.kanjiquest.core.srs.SrsAlgorithm
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class StudyCard(
    val kanji: Kanji,
    val vocabulary: List<Vocabulary>,
    val srsCard: SrsCard?
)

data class FlashcardStudyUiState(
    val cards: List<StudyCard> = emptyList(),
    val currentIndex: Int = 0,
    val isFlipped: Boolean = false,
    val isComplete: Boolean = false,
    val isLoading: Boolean = true,
    val totalStudied: Int = 0,
    val gradeResults: Map<Int, Int> = emptyMap()
) {
    val currentCard: StudyCard? get() = cards.getOrNull(currentIndex)
    val progress: String get() = "${currentIndex + 1} of ${cards.size}"
}

enum class StudyGrade(val quality: Int, val label: String) {
    AGAIN(1, "Again"),
    HARD(2, "Hard"),
    GOOD(4, "Good"),
    EASY(5, "Easy")
}

@HiltViewModel
class FlashcardStudyViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val flashcardRepository: FlashcardRepository,
    private val kanjiRepository: KanjiRepository,
    private val srsRepository: SrsRepository,
    private val srsAlgorithm: SrsAlgorithm
) : ViewModel() {

    private val deckId: Long = checkNotNull(savedStateHandle["deckId"])

    private val _uiState = MutableStateFlow(FlashcardStudyUiState())
    val uiState: StateFlow<FlashcardStudyUiState> = _uiState.asStateFlow()

    private val gradeResults = mutableMapOf<Int, Int>()

    init {
        loadCards()
    }

    private fun loadCards() {
        viewModelScope.launch {
            val kanjiIds = flashcardRepository.getKanjiIdsByDeck(deckId)
            val cards = kanjiIds.mapNotNull { id ->
                val kanji = kanjiRepository.getKanjiById(id) ?: return@mapNotNull null
                val vocab = kanjiRepository.getVocabularyForKanji(id).take(3)
                val srsCard = srsRepository.getCard(id)
                StudyCard(kanji, vocab, srsCard)
            }.shuffled()

            _uiState.value = FlashcardStudyUiState(
                cards = cards,
                isLoading = false
            )
        }
    }

    fun flip() {
        _uiState.value = _uiState.value.copy(isFlipped = true)
    }

    fun grade(grade: StudyGrade) {
        val current = _uiState.value.currentCard ?: return
        val kanjiId = current.kanji.id

        gradeResults[kanjiId] = grade.quality

        viewModelScope.launch {
            // Update SRS card
            val card = srsRepository.getCard(kanjiId)
            if (card != null) {
                val currentTime = kotlinx.datetime.Clock.System.now().epochSeconds
                val updated = srsAlgorithm.review(card, grade.quality, currentTime)
                srsRepository.saveCard(updated)
            }

            // Update flashcard study count
            flashcardRepository.markStudied(deckId, kanjiId)

            // Advance to next card or complete
            val nextIndex = _uiState.value.currentIndex + 1
            if (nextIndex >= _uiState.value.cards.size) {
                _uiState.value = _uiState.value.copy(
                    isComplete = true,
                    totalStudied = _uiState.value.cards.size,
                    gradeResults = gradeResults.toMap()
                )
            } else {
                _uiState.value = _uiState.value.copy(
                    currentIndex = nextIndex,
                    isFlipped = false
                )
            }
        }
    }
}

File: ios-app/KanjiQuest/Screens/Flashcard/FlashcardStudyViewModel.swift
import Foundation
import SharedCore

struct StudyCard: Identifiable {
    var id: Int32 { kanji.id }
    let kanji: Kanji
    let vocabulary: [Vocabulary]
    let srsCard: SrsCard?
}

enum StudyGrade: Int, CaseIterable {
    case again = 1
    case hard = 2
    case good = 4
    case easy = 5

    var label: String {
        switch self {
        case .again: return "Again"
        case .hard: return "Hard"
        case .good: return "Good"
        case .easy: return "Easy"
        }
    }
}

@MainActor
final class FlashcardStudyViewModel: ObservableObject {
    @Published var cards: [StudyCard] = []
    @Published var currentIndex = 0
    @Published var isFlipped = false
    @Published var isComplete = false
    @Published var isLoading = true
    @Published var totalStudied = 0
    @Published var gradeResults: [StudyGrade: Int] = [:]

    private var flashcardRepository: FlashcardRepository?
    private var kanjiRepository: KanjiRepository?
    private var srsRepository: SrsRepository?
    private var srsAlgorithm: Sm2Algorithm?
    private var deckId: Int64 = 1

    func load(container: AppContainer, deckId: Int64) {
        flashcardRepository = container.flashcardRepository
        kanjiRepository = container.kanjiRepository
        srsRepository = container.srsRepository
        srsAlgorithm = container.srsAlgorithm
        self.deckId = deckId

        Task {
            let kanjiIds = (try? await flashcardRepository?.getKanjiIdsByDeck(deckId: deckId)) ?? []
            var studyCards: [StudyCard] = []
            for kanjiId in kanjiIds {
                if let kanji = try? await kanjiRepository?.getKanjiById(id: Int32(kanjiId)) {
                    let vocab = (try? await kanjiRepository?.getVocabularyForKanji(kanjiId: Int32(kanjiId))) ?? []
                    let srsCard = try? await srsRepository?.getCard(kanjiId: Int32(kanjiId))
                    studyCards.append(StudyCard(kanji: kanji, vocabulary: Array(vocab.prefix(3)), srsCard: srsCard))
                }
            }
            cards = studyCards.shuffled()
            isLoading = false
        }
    }

    func flip() {
        isFlipped = true
    }

    func grade(_ grade: StudyGrade) {
        gradeResults[grade, default: 0] += 1
        totalStudied += 1

        if currentIndex < cards.count {
            let card = cards[currentIndex]
            Task {
                // Update SRS card via algorithm
                if let existingCard = try? await srsRepository?.getCard(kanjiId: card.kanji.id) {
                    let currentTime = Int64(Date().timeIntervalSince1970)
                    if let updated = srsAlgorithm?.review(card: existingCard, quality: Int32(grade.rawValue), currentTime: currentTime) {
                        try? await srsRepository?.saveCard(card: updated)
                    }
                }

                // Mark studied in flashcard deck
                try? await flashcardRepository?.markStudied(deckId: deckId, kanjiId: card.kanji.id)
            }
        }

        // Advance to next card
        isFlipped = false
        if currentIndex + 1 < cards.count {
            currentIndex += 1
        } else {
            isComplete = true
        }
    }

    var currentCard: StudyCard? {
        guard currentIndex < cards.count else { return nil }
        return cards[currentIndex]
    }
}
